lib/graphql/dataloader/source.rb

```diff
@@ -35,24 +32,23 @@ def request_all(keys)
       # @return [Object] The result from {#fetch} for `key`. If `key` hasn't been loaded yet, the Fiber will yield until it's loaded.
       def load(key)
         if @results.key?(key)
-          @results[key]
+          result_for(key)
         else
           @pending_keys << key
           sync
-          @results[key]
+          result_for(key)
         end
       end
 
       # @param keys [Array<Object>] Loading keys which will be passed to `#fetch` (or read from the internal cache).
       # @return [Object] The result from {#fetch} for `keys`. If `keys` haven't been loaded yet, the Fiber will yield until they're loaded.
       def load_all(keys)
-        if keys.any? { |k| !@results.key?(k) }
-          pending_keys = keys.select { |k| !@results.key?(k) }
+        if pending_keys = keys.select { |k| !@results.key?(k) }
```

Comment:
Evidently there's no test for this, but I think this will not have the desired effect when all of `keys` are present in `@results`. For example: 

```
results = { a: 1, b: 2, c: 3 }
keys = [:a, :b, :c]
if pending_keys = keys.select { |k| !results.key?(k) }
  puts "has pending keys: #{pending_keys}"
else
  puts "has no pending keys"
end

# $ ruby pending_test.rb
# has pending keys: []
```

Even when all `keys` are filtered out, `pending_keys` is `[]`, which is _truthy_, so it would enter the `if ...` block here. 

Sorry there's no test for it, but how about either: 

- Adding a test to demonstrate that the current implementation isn't subject to the bug I described above 
- Or, reverting the implementation to the previous one 

? 

Alternatively, if you really want to skip the extra Array, there probably _is_ a zero-allocation approach, something like 

```ruby 
def load_all(keys)
  any_pending = false 
  
  keys.each do |k|
    if !@results.key?(k)
      any_pending = true 
      @pending_keys << k 
    end 
  end 
  
  if any_pending 
    sync 
  end 
  
  @keys.map { |k| result_for(k) }
end 
```

That'd be fine with me too. But I want to make sure we don't `sync` when all of the given `keys` are already present in `@results`.

Reply:
Yeah, you're right. Sorry about that. Feels like more effort than it's worth so I reverted the change.